Improve texture modifier documentation (#16753)
minetest/minetest: doc/lua_api.md

#### Crack

* `[crack:<tilecount>:<frame>`
* `[cracko:<tilecount>:<frame>`
* `[crack:<framecount>:<tilecount>:<frame>`
* `[cracko:<framecount>:<tilecount>:<frame>`

Overlay a scaled frame of the crack texture, `crack_anylength.png`,
over a texture, with options for alpha and blitting on all frames.

With `crack`, the crack will be overlaid over the whole base
texture, while with `cracko`, the crack will only be overlaid
over fully opaque base texture regions.

* `framecount`: Optional. Vertical animation frame count of the crack texture;
  usually omitted.
* `tilecount`: Vertical and horizontal tile count of the base texture. The crack
 will be blit on each tile of the base texture. Usually `1`.
* `frame`: Current animation frame ("crack progression").

Note: This always scales the crack to the size of the base texture (or the tiles

> of the base texture, if tilesize is provided).

Author: I've completely rewritten the crack section.

Examples:

    default_stone.png^[crack:1:2

* Cracked stone
* Use the third crack progression (`2` because of 0-indexing) ...
* ... and draw it on a stone texture consisting of `1` tile

>     default_lava_source_animated.png^[crack:1:8:2

Author: No, this is correct. But the leading `1` can be omitted.

* Cracked lava
* The base texture has 8 animated vertical frames ...
* ... and exactly `1` "tile" horizontally (frames are not tiles)

#### `[combine:<w>x<h>:<textures>

* `<w>`: width of the resulting texture
* `<h>`: height of the resulting texture
* `<textures>: Colon (`:`)-separated list of locations `x`, `y` (both positive

>   integers) and escaped texture modifiers texture to blit in the form

Reviewer: `escaped texture modifiers texture` isn't clear. Perhaps add something like `<texture> can contain texture modifiers, but these must be escaped according to the rules in the Escaping section above` below the next line.

Or introduce the term in the introduction?

    default_sandstone.png^[resize:16x16

#### `[opacity:<ratio>`

Makes the base image transparent according to the given `ratio`.

This multiplies the alpha value of each pixel of the base texture
with `ratio/255` and rounds properly afterwards.

`ratio` must be between 0 (transparent) and 255 (opaque).

Example:

    default_sandstone.png^[opacity:127

#### `[invert:<mode>`

* `mode` is a string which may contain the characters `r`, `g`, `b` and `a`.

The channels corresponding to the occurring characters (red, green, blue
and alpha) will be inverted (set to `255-value`).

Example:

    default_apple.png^[invert:rgb

#### `[brighten`

Interpolates 50:50 between the color of each pixel of the base texture and
white.

Example:

    tnt_tnt_side.png^[brighten

#### `[noalpha`

Makes the texture completely opaque by setting the alpha channel of the base
texture to the maximum value (255).

As the red, green and blue channels aren't premultiplied with alpha in PNGs,

> this might reveal hidden colors of otherwise transparent portions of an image.

Reviewer: I think this affects other operations like `[invert`, too. Perhaps make this a general comment with a suitable link to e.g. https://en.wikipedia.org/wiki/Alpha_compositing (not sure if that is a good resource, though).

Convert one color to transparency.

Pixels of the base texture having the exact same RGB color will have their
alpha value set to 0.

As the red, green and blue channels are kept, the original color can be
restored using `[noalpha` (which will however also make originally
semi-transparent portions of the image opaque).

Example:

    default_cobble.png^[makealpha:128,128,128

#### `[transform<t>`

Rotates and/or flips the image.

> `<t>` is the concatenation of either numbers or names identifying

Author: Done. But all the number/integer confusion is bound to be fixed in #16735 before.

#### Crack

* `[crack:<vertical_frames>:<crack_frame>`
* `[cracko:<vertical_frames>:<crack_frame>`
* `[crack:<grid>:<vertical_frames>:<crack_frame>`
* `[cracko:<grid>:<vertical_frames>:<crack_frame>`

This draws a step of the crack animation on the texture.

It overlays a scaled frame of the crack texture, `crack_anylength.png`,
over a texture, with options for alpha and blitting multiple times for
tile sheets / animated textures.

With `crack`, the crack will be overlaid over the whole base

> texture, while with `cracko`, the crack will only be overlaid

Reviewer 2: ```suggestion
With `[crack`, the crack will be overlaid over the whole base
texture, while with `[cracko`, the crack will only be overlaid
```

.. perhaps, to be more precise?

over fully opaque base texture regions.

> Parameters if `grid` is omitted:

Reviewer 2: Does this need special treatment? Just specify that `grid` defaults to `1` if not specified. The subsequent explanations would then still hold true. Avoids duplication.

* `vertical_frames`: Amount of times the crack will be applied
  in vertical direction on the base texture (integer [imagesize])
* `crack_frame`: Crack animation frame (integer [imageframe], counting
  starts at 0)

Parameters if `grid` is present:

* `grid`: Creates a square grid of multiple cracks, with
  side length `grid`, e.g. the number 3 creates a 3x3 grid of cracks
  (integer [imagesize], default 1)
* `vertical_frames`: Amount of times the grid created by `grid`
  will be repeated vertically (integer [imagesize], often `1`)
* `crack_frame`: (see above)

Note: The resulting crack (or cracks) texture will always be scaled
to the size of the base texture.

Examples:

    default_stone.png^[crack:1:2

* Cracked stone
* Use the third crack progression (`2` because of 0-indexing)
  and draw it once on a stone texture

    default_lava_source_animated.png^[crack:8:2

* Cracked lava
* Use the 3rd crack progression and draw it 8 times on top of the base texture
* The base texture should have 8 vertical frames as well

    example.png^[crack:3:1:0

* Blits a 3x3 grid of crack textures, each at frame 0, on `example.png`

> #### `[combine:<w>x<h>:<textures>`

Reviewer 2: I propose to name `textures`  `parts`, `sections` or `insertions` because `textures` is not just concat' texture names

Reviewer 3: `parts` sounds good to me. maybe `blits`?

#### Crack

* `[crack:<vertical_frames>:<crack_frame>`
* `[cracko:<vertical_frames>:<crack_frame>`
* `[crack:<grid>:<vertical_frames>:<crack_frame>`
* `[cracko:<grid>:<vertical_frames>:<crack_frame>`

This draws a step of the crack animation on the texture.

> More precisely, it overlays a scaled frame of the crack texture,

Reviewer 4: ```suggestion
It overlays a scaled frame of the crack texture,
```

`crack_anylength.png`, over a texture, with options for alpha and blitting
multiple times for tile sheets / animated textures.

With `crack`, the crack will be overlaid over the whole base
texture, while with `cracko`, the crack will only be overlaid
over fully opaque base texture regions.

Parameters if `grid` is omitted:

* `vertical_frames`: Amount of times the crack will be blit on the

>   base texture, stacked vertically (integer [imagesize])

Reviewer 4: This sounds confusing to me. Can't you just write something like
Amount of times the crack frame will be applied repeatedly in vertical direction.

* `crack_frame`: Crack animation frame (integer [imageframe], counting
  starts at 0)

Parameters if `grid` is present:

* `grid`: Creates a square grid of multiple cracks, with
  side length `grid`, e.g. the number 3 creates a 3x3 grid of cracks
  (integer [imagesize], default 1)

> * `vertical_frames`: Amount of times the grid created by `tile_count`

Reviewer 4: What does ""created by `tile_count`" mean here?
I can't find the term `tile_count` anywere.

Author: It should have been `grid`.

  will be repeated vertically (integer [imagesize], often `1`)
* `crack_frame`: (see above)

Note: The resulting crack (or cracks) texture will always be scaled
to the size of the base texture.

Examples:

    default_stone.png^[crack:1:2

* Cracked stone
* Use the third crack progression (`2` because of 0-indexing) ...

> * ... and draw it once on a stone texture

Reviewer 4: ```suggestion
* Use the third crack progression (`2` because of 0-indexing)
  and draw it once on a stone texture
```

    default_lava_source_animated.png^[crack:8:2

* Cracked lava
* Use the 3rd crack progression ...

> * ... and draw it 8 times on top of the base texture

Reviewer 4: ```suggestion
* Use the 3rd crack progression and draw it 8 times on top of the base texture
```

* The base texture should have 8 vertical frames as well

    example.png^[crack:3:1:0

* Blits a 3x3 grid of crack textures, each at frame 0, on `example.png`

#### `[combine:<w>x<h>:<textures>`

* `<w>`: width of resulting texture (integer [imagesize])
* `<h>`: height of resulting texture (integer [imagesize])

> * `<textures>: Colon-(`:`)-separated list of locations `x`, `y` and textures to

Reviewer 4: ```suggestion
* `<textures>`: Colon-separated (`:`) list of locations `x`, `y` and textures to
```

  blit; written in the form `<x>,<y>=<texture>` for each texture. Can be empty.

`x` and `y` are integers [imageframe].
A `<texture>` (in `<textures>`) can contain texture modifiers, but these must

> be escaped according to the rules in the Escaping section above.

Reviewer 4: ```suggestion
be escaped according to the rules in [Escaping](#Escaping).
```

    default_sandstone.png^[resize:16x16

#### `[opacity:<ratio>`

Makes the base image transparent according to the given `ratio`.

This multiplies the alpha value of each pixel of the base texture
with `ratio/255` and rounds the result afterwards (using the

> "round half up" rule).

Reviewer 4: ```suggestion
with `ratio/255` and rounds the result to the closest integer afterwards.
```

    default_sandstone.png^[resize:16x16

#### `[opacity:<ratio>`

Makes the base image transparent according to the given `ratio`.

This multiplies the alpha value of each pixel of the base texture

> with `ratio/255` and rounds properly afterwards.

Reviewer 4: what does "properly" mean? up? down?

Author: It's the "round half up rule". Fixed.

`ratio` is an integer in range [0, 255]. 0 = transparent, 255 = opaque.

Example:

    default_sandstone.png^[opacity:127

#### `[invert:<mode>`

* `mode` is a string which may contain the characters `r`, `g`, `b` and `a`.

The channels corresponding to the occurring characters (red, green, blue

> and alpha) will be inverted (set to `255-value`).

Reviewer 4: ```suggestion
and alpha) will be inverted (set to `255 - value`).
```

Example:

    default_apple.png^[invert:rgb

#### `[brighten`

Interpolates 50:50 between the color of each pixel of the base texture and
white.

Example:

    tnt_tnt_side.png^[brighten

#### `[noalpha`

Makes the texture completely opaque by setting the alpha channel of the base

> texture to the maximum value (255).

Reviewer 4: ```suggestion
texture to the maximum value (`255`).
```

> As the red, green and blue channels aren't premultiplied with alpha in PNGs,

Reviewer 4: ```suggestion
As the red, green and blue channels aren't pre-multiplied with alpha in PNGs,
```

Convert the given color to transparency.

Pixels of the base texture having the exact same RGB color will have their

> alpha value set to 0.

Reviewer 4: ```suggestion
alpha value set to `0`.
```

#### `[transform<t>`

Rotates and/or flips the image.

`<t>` is the concatenation of either integers or names identifying
transformations from the following table:

    0  I      identity (no transformation)
    1  R90    rotate by 90° counterclockwise
    2  R180   rotate by 180° counterclockwise
    3  R270   rotate by 270° counterclockwise
    4  FX     flip X
    5  FXR90  flip X then rotate by 90° counterclockwise
    6  FY     flip Y
    7  FYR90  flip Y then rotate by 90° counterclockwise

Transformation names are case-insensitive.

Example:

    default_stone.png^[transformFXR90

#### `[inventorycube{<top>{<left>{<right>`

Renders a cube similar to how it would look in the inventory with the three
given textures using simple software rendering.

The resulting image will be 9 times the nearest power of 2 large, enough to
contain the dimensions of the largest image, clamped to a range of at least

> 4 and at most 64.

Reviewer 2: Do we really want to document this? This seems oddly specific and I'd rather not want modders to rely on this.

Reviewer: If one wants to overlay something on the resulting image or combine it with other things it is unfortunately necessary to know the exact size of the result, because currently most modifiers take pixel coordinates.

It would be really great to have a size agnostic modifier language, because different clients may be using different texture packs with different resolutions. Something like accepting coordinates with a decimal point as relative, e.g. `[resize:1.x2.]` doubling height.

Author: Yeah, I have to agree with @j-r here, the size needs to be documened.

If you want to reserve the right to arbitrarily change the output size of this element in future: Please don’t, you will just break mods that assume the current size (regardless of whether it is documentated or not).

If you *want* to change the output size in future, you can just add more optional parameters to this element (with the current hardcoded size as default), so backwards-compatibility is preserved.

Reviewer 3: Perhaps we should add an option for specifying a target resolution of `[inventorycube`.

I don't think there's any harm in keeping the current formula though. Maybe add a comment to the code that it should not be changed.

"Layouting" capabilities / resolution-agnostic texture modifiers more generally are a trickier discussion for another issue (related: my opinions on texture modifier design #11588), and also ties into SSCSM: The most powerful solution is obviously if you can do whatever image manipulation you desire (#16794).

Reviewer 2: The current formula (= the returned image) depends on the node texture size, hence I doubt this information is of any use to modders because the resolution cannot be retrieved in any helpful way. `[resize` will always be necessary to cover all cases.

Reviewer 3: I don't think it's terribly important to document this. Most (all?) modders indeed won't get any use out of it. At the same time, I think there's no point in ever changing the current formula as it is. Indeed I think it's likely that people have come to rely on it, if only purely by experimentation (which probably means their code implicitly relies on, say, 16x16 node textures).

What might be important to mention still is that `[inventorycube` resolution grows with the node texture resolution, ensuring that `[inventorycube` is crisp (and implying that modders expecting a fixed resolution should `[resize`).

Author: Simplified the description.

Reviewer: > The current formula (= the returned image) depends on the node texture size, hence I doubt this information is of any use to modders because the resolution cannot be retrieved in any helpful way. `[resize` will always be necessary to cover all cases.

Yes, but if the input texture has already been resized to a known size to apply some size dependent modifier then it should not be necessary to resize the output of `[inventorycube` again. (Though I haven't tested whether complex input textures actually work with the special escaping here, but it should, I think.)

#### `[transform<t>`

Rotates and/or flips the image.

`<t>` is the concatenation of either integers or names identifying
transformations from the following table:

    0  I      identity (no transformation)
    1  R90    rotate by 90° counterclockwise
    2  R180   rotate by 180° counterclockwise
    3  R270   rotate by 270° counterclockwise
    4  FX     flip X
    5  FXR90  flip X then rotate by 90° counterclockwise
    6  FY     flip Y
    7  FYR90  flip Y then rotate by 90° counterclockwise

Transformation names are case-insensitive.

Example:

    default_stone.png^[transformFXR90

#### `[inventorycube{<top>{<left>{<right>`

Renders a cube similar to how it would look in the inventory with the three
given textures using simple software rendering.

The resulting image will be 9 times the nearest power of 2 large, enough to
contain the dimensions of the largest image, clamped to a range of at least
4 and at most 64.

As a formula:

>     9 * max(4, min(64, 2^ceil(log_2(max(d)))))

Reviewer 4: ```suggestion
    `9 * max(4, min(64, 2^ceil(log_2(max(d)))))`
```

Author: No. In Markdown, indenting text with 4 spaces makes it a code block. https://daringfireball.net/projects/markdown/syntax#precode

where `d` is the set of dimensions (width and height) of all faces.

Escaping does not apply here and `^` is replaced by `&` in texture names
instead.

Example:

    [inventorycube{grass.png{dirt.png&grass_side.png{dirt.png&grass_side.png

> Creates a cube with `grass.png`, `dirt.png^grass_side.png` and

Reviewer 4: It does not actually create a cube so inventorycube is more correct here.

Author: I’ll go with “Renders a cube” cuz “inventorycube” is a strange word.

    [fill:16x16:#20F02080
    texture.png^[fill:8x8:4,4:red

> #### `[lowpart:<percent>:<texture>`

Reviewer 3: shouldn't there be a "\* `texture` is an escaped texture modifier" here?

    default_torch_animated.png^[verticalframe:16:8

#### `[mask:<texture>`

* `texture` is an escaped texture modifier

Applies a *bitwise and* to all RGBA values of `texture` and the base texture.

If a pixel of the base texture is out of bounds on texture, it is preserved.

> Masking is associative and commutative if all involved textures have the same

Reviewer 3: not sure why i found this important to mention at the time, i think this can be omitted. it's clear from the definition.

dimensions.

*See notes: `TEXMOD_UPSCALE`*

#### `[sheet:<w>x<h>:<x>,<y>`

* `<w>`: sheet width in tiles (integer [imagesize])
* `<h>`: sheet height in tiles (integer [imagesize])
* `<x>`: x position in tiles (integer [imageframe], starts at 0)
* `<y>`: y position in tiles (integer [imageframe], starts at 0)

Retrieves the tile of a base image at position `x`, `y`,
which is assumed to be a tilesheet with dimensions
`w`, `h` (in tiles).

#### `[colorize:<color>:<ratio>`

> 

Reviewer 3: ```suggestion
```

* `color` is a ColorString
* `ratio` is an optional integer in range [0, 255] or the string `"alpha"`

Colorizes the textures with the given color.

> Interpolates between `color` and the pixel colors of the base texture as specified

Reviewer 3: I think we need to be careful here. If alpha were interpolated, we would expect `[colorize` to make images semitransparent normally, which we don't want. What exactly happens with `alpha`?

Peeking at the implementation (`apply_colorize` in `imagesource.cpp`): Something funny.

Let $r$ be the "effective ratio" (either the given number, or the alpha channel of the color if none is given).

In all cases, only pixels with alpha > 0 are touched.

If we have a ratio that is not alpha, or if the ratio is alpha but the alpha of the color is less than 255, we get an interpolation that *includes alpha*. I don't think this makes much sense for actual semitransparent pixels but that's how it is. It follows that the only real sensible usage we should recommend in this case is with a base image that does not use semitransparency, and a color that has an alpha of 255.

If the ratio is alpha, *and* the color has an alpha of 255, the alpha channel of the original image is kept, but the color replaced.

This seems quite odd to me. Maybe I'm misreading something, maybe there's a bug, maybe the documentation needs to be fixed. In any case I think something funky is going on, and I wouldn't bet on myself from some years back having analyzed this properly.

Author: I would say this behavior is an outright bug. I have tested this in a test mod and can confirm what you said.

The point of `colorize` is to change the color of a base image. That’s it. That was always it’s “selling point”. Therefore, it should not randomly modify alpha IMHO. I think a functional `colorize` implementation should guarantee that the alpha values of the output image are always the same as from the input image. Especially the fact that the "funky" behavior is only triggered for very specific values is pretty sus.

Exception: When the `"alpha"` ratio is used. I noticed that `"alpha"` makes no sense when you also keep the alpha channel in the output image. The lower the alpha, the lesser visible the lack of color change becomes. Not very useful. What I think was intended here is that this parameter means, that the alpha channel of the base image is only used as a separate source of information, not also as actual base image input. The "true" base image would thus only be the raw color or it, assuming alpha of 255 everywhere, so the output image is always fully opaque.

In any case, for the documentation I’ve decided to add several warnings and guide the modder to only use specific values for now. In the long runs, `[colorize` needs to be fixed IMHO.

by the `ratio`:

* Defaults to the alpha of `color` if omitted
* If it's an integer from 0 (only base texture color) to 255 (only `color`),
  it's directly used as interpolation ratio: The resulting color of a pixel is

>   `ratio` times `color` plus `(255 - ratio)` times base texture color;

Reviewer 2: ```suggestion
* If it's an integer from 0 (only base texture color) to 255 (only `color`),
  the resulting color of a pixel is `ratio * color + (255 - ratio) * base_tex_color`.
```

* If it's the string `"alpha"`, the texture pixel's alpha value determines the

>   ratio per pixel

Reviewer 2: I'd prefer to clearly mention `ratio` here because "alpha" is most likely associated with "color" and not "ratio".
```suggestion
* If `ratio = "alpha"`, the texture pixel's alpha value determines the
  ratio per pixel
```

    default_torch_animated.png^[verticalframe:16:8

#### `[mask:<texture>`

* `texture` is an escaped texture modifier

Applies a *bitwise and* to all RGBA values of `texture` and the base texture.

If a pixel of the base texture is out of bounds on texture, it is preserved.

Masking is associative and commutative if all involved textures have the same
dimensions.

*See notes: `TEXMOD_UPSCALE`*

#### `[sheet:<w>x<h>:<x>,<y>`

* `<w>`: sheet width in tiles (integer [imagesize])
* `<h>`: sheet height in tiles (integer [imagesize])
* `<x>`: x position in tiles (integer [imageframe], starts at 0)
* `<y>`: y position in tiles (integer [imageframe], starts at 0)

Retrieves the tile of a base image at position `x`, `y`,
which is assumed to be a tilesheet with dimensions
`w`, `h` (in tiles).

#### `[colorize:<color>:<ratio>`

* `color` is a ColorString
* `ratio` is an optional integer in range [0, 255] or the string `"alpha"`

Colorizes the textures with the given color.

> Interpolates between color and the pixel colors of the base texture as specified

Reviewer 4: ```suggestion
Interpolates between `color` and the pixel colors of the base texture as specified
```
Here too?

    default_torch_animated.png^[verticalframe:16:8

#### `[mask:<texture>`

* `texture` is an escaped texture modifier

> Applies a *bitwise and* (`bit.band`) to all RGBA values of `texture` and

Reviewer 4: ```suggestion
Applies a *bitwise and* to all RGBA values of `texture` and
```
It doesn't use `bit.band`.

the base texture.

If a pixel of the base texture is out of bounds on texture, it is preserved.

Masking is associative and commutative if all involved textures have the same
dimensions.

*See notes: `TEXMOD_UPSCALE`*

#### `[sheet:<w>x<h>:<x>,<y>`

* `<w>`: sheet width in tiles (integer [imagesize])
* `<h>`: sheet height in tiles (integer [imagesize])
* `<x>`: x position in tiles (integer [imageframe], starts at 0)
* `<y>`: y position in tiles (integer [imageframe], starts at 0)

Retrieves the tile of a base image at position `x`, `y`,
which is assumed to be a tilesheet with dimensions
`w`, `h` (in tiles).

#### `[colorize:<color>:<ratio>`

* `color` is a ColorString
* `ratio` is an optional integer in range [0, 255] or the string "alpha"

Colorizes the textures with the given color.

Interpolates between color and the pixel colors of the base texture as specified
by the `ratio`:

> * Defaults to the alpha of `color` if omitted;

Reviewer 4: ```suggestion
* Defaults to the alpha of `color` if omitted
```

> * If it's an integer from 0 (only pixel color) to 255 (only color), it is

Reviewer 4: ```suggestion
* If it's an integer from 0 (only pixel color) to 255 (only `color`), it is
```

Author: No, here "color" is meant as the English word.

Reviewer 4: I thought this would make it clearer, since "color" here refers to the specified `color`, right?
Otherwise, the distinction between " only pixel color" and "only color" may be harder to understand.

  directly used as interpolation ratio: The resulting color of a pixel is

>   ratio times color plus `(255-ratio)` times pixel color;

Reviewer 4: ```suggestion
  `ratio` times `color` plus `255 - ratio` times pixel color
```

> * If it's the string "alpha", the texture pixel's alpha value determines the

Reviewer 4: ```suggestion
* If it's the string `"alpha"`, the texture pixel's alpha value determines the
```

#### `[colorize:<color>:<ratio>`

* `color` is a ColorString (should not use alpha)
* `ratio` is an optional integer in range [0, 255] or the string `"alpha"`

Colorizes the textures with the given color.

Interpolates between `color` and the pixel colors of the base texture as specified
by the `ratio`:

* Defaults to the alpha of `color` if omitted (NOT RECOMMENDED)
* If it's an integer from 0 (only base texture color) to 255 (only `color`),
  the resulting color of a pixel is `ratio * color + (255 - ratio) * base_tex_color`.
* If `ratio = "alpha"`, the texture pixel's alpha value determines the
  ratio per pixel (NOT RECOMMENDED)

> * NOTE: Due to bugs, we currently can only recommend the base image to be fully

Reviewer 3: 👍 on the "not recommended's" *but* I propose this be relaxed to "we currently can only recommend the base image to **avoid any semitransparent pixels** (alpha > 0, < 255)".

alpha = 0 is fine and will continue to work as before when we fix this

Reviewer 3: https://github.com/luanti-org/luanti/pull/16753/changes/b1c1f770adf9861e4fb9da0cfc693f4d9113399b

#### `[multiply:<color>`

Multiplies the RGB values of the base texture per pixel with the RGB

> values of color; the `alpha` value of `color` is ignored.

Reviewer 4: ```suggestion
values of `color`; the `alpha` value of `color` is ignored.
```
